src/lib.rs

```diff
@@ -66,222 +36,460 @@ pub fn invoke(params: u32) -> u32 {
-                RawBytes::serialize(FundParams {
-                    value: stake.clone(),
-                })?,
-                TokenAmount::zero(),
-            )?;
+    // just a test address
+    const IPC_GATEWAY_ADDR: u64 = 1024;
+
+    fn std_construct_param() -> ConstructParams {
+        ConstructParams {
+            parent: Default::default(),
+            name: "".to_string(),
+            ipc_gateway_addr: IPC_GATEWAY_ADDR,
+            consensus: ConsensusType::Delegated,
+            min_validator_stake: Default::default(),
+            min_validators: 0,
+            finality_threshold: 0,
+            check_period: 0,
+            genesis: vec![],
         }
+    }
 
-        // remove stake from balance table
-        st.rm_stake(&caller, &stake)?;
+    fn construct_runtime() -> MockRuntime {
+        let caller = *INIT_ACTOR_ADDR;
+        let receiver = Address::new_id(1);
+        let mut runtime = MockRuntime::new(caller, receiver);
 
-        // send back to owner
-        st.send(&caller, METHOD_SEND, RawBytes::default(), stake)?;
+        let params = std_construct_param();
 
-        st.mutate_state();
-        st.save();
-        Ok(None)
+        runtime.expect_validate_caller_addr(vec![caller]);
+
+        runtime
+            .call::<Actor>(
+                Method::Constructor as u64,
+                &cbor::serialize(&params, "test").unwrap(),
+            )
+            .unwrap();
+
+        runtime
     }
 
-    fn kill() -> anyhow::Result<Option<RawBytes>> {
-        let mut st = State::load();
+    #[test]
+    fn test_constructor() {
+        let runtime = construct_runtime();
+        assert_eq!(runtime.state.is_some(), true);
+        // let state: State = runtime.state().unwrap();
+    }
 
-        if st.status == Status::Terminating || st.status == Status::Killed {
-            abort!(
-                USR_ILLEGAL_STATE,
-                "the subnet is already in a killed or terminating state"
-            );
-        }
-        if st.validator_set.len() != 0 {
-            abort!(
-                USR_ILLEGAL_STATE,
-                "this subnet can only be killed when all validators have left"
-            );
-        }
+    #[test]
+    fn test_join_fail_no_min_collateral() {
+        let mut runtime = construct_runtime();
 
-        // move to terminating state
-        st.status = Status::Terminating;
+        let validator = Address::new_id(100);
+        let params = JoinParams {
+            validator_net_addr: validator.to_string(),
+        };
 
-        // unregister subnet
-        st.send(
-            &Address::new_id(ext::sca::SCA_ACTOR_ADDR),
-            Method::Kill as u64,
-            RawBytes::default(),
-            TokenAmount::zero(),
-        )?;
+        let r = runtime.call::<Actor>(
+            Method::Join as u64,
+            &cbor::serialize(&params, "test").unwrap(),
+        );
 
-        st.mutate_state();
-        st.save();
-        Ok(None)
+        assert_eq!(
```

Comment:
You can check call failures like here: https://github.com/consensus-shipyard/builtin-actors/blob/fa871a816314aa7ebbb908c8fd6aca1b6f62796d/actors/hierarchical_sca/tests/harness/mod.rs#L128-L135